### PR TITLE
Add future sessions to weekend view.

### PR DESCRIPTION
**Pull request by bashopman**
Implements #67 .

### CODE REVIEW

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 1]
```diff
@@ -117,12 +151,60 @@ class F1TvClient @Inject constructor(
...
+                        }
+                }
+            return schedules.map {
+                F1TvSession(
+                    id = F1TvSessionId(it.id),
+                    eventId = event.id,
+                    pictureUrl = PICTURE_URL.format(it.metadata.pictureUrl),
+                    contentId = it.metadata.contentId,
+                    name = it.metadata.title,
+                    contentSubtype = dateTimeFormatter.format(Instant.ofEpochMilli(it.metadata.emfAttributes.sessionStartDate)),
```
Comment (bashopman) [inline app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:196]: I'm abusing the `contentSubType` to allow showing the start time of the future session. What is the better way to do this?

**Comment by bashopman** [comment]
Result:
image
@leonardoxh please take a look. And if you have a suggestion to avoid the mis-use of the `contentSubtype` field, let me know.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt (hunk 1 above)
Comment (leonardoxh) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:196]: Well, if we had something better in place I'd say to change, but I think for now at least is ok.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 2]
```diff
@@ -142,31 +224,32 @@ class F1TvClient @Inject constructor(
...
+                            driver = F1TvDriverId("") //TODO - do we have to load the driver ?
+                        )
+                    } else {
+                        F1TvBasicChannel(
+                            channelId = channelIdAndContentId.first().split("=").last(),
+                            contentId = channelIdAndContentId.last().split("=").last(),
+                            type = F1TvBasicChannelType.from(it.type, it.title)
+                        )
+                    }
+                } ?: listOf()
```
Comment (leonardoxh) [inline app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:252]: I think this whole change can be reverted, did you use a auto formatter?

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/session.kt [hunk 3]
```diff
@@ -50,3 +50,58 @@ data class F1TvSession(
...
+    val contentSubtype: String,
+    val contentId: String
+)
+
+@JsonClass(generateAdapter = true)
+data class F1TvFutureSessionEmfAttributes(
+    val sessionStartDate: Long,
+    val sessionEndDate: Long,
+    @Json(name = "Series") val series: String,
+
```
Comment (leonardoxh) [inline app/src/main/java/fr/groggy/racecontrol/tv/f1tv/session.kt:106]: revert this whitespace.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/session.kt [hunk 4]
```diff
@@ -50,3 +50,58 @@ data class F1TvSession(
...
+    )
```
Comment (leonardoxh) [inline app/src/main/java/fr/groggy/racecontrol/tv/f1tv/session.kt:107]: Identation here

File: app/src/main/res/values-nl/strings.xml [hunk 5]
```diff
@@ -14,7 +14,7 @@
         wisselen. Om het op te lossen moet je contact opnemen met F1 ondersteuning.
     </string>
 
-    <string name="unable_to_play_video_message">Er ging iets fout bij het afspelen van deze video</string>
+    <string name="unable_to_play_video_message">Er ging iets fout bij het afspelen van deze video...</string>
```
Comment (leonardoxh) [inline app/src/main/res/values-nl/strings.xml:17]: revert

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 6]
```diff
@@ -95,10 +117,22 @@ class F1TvClient @Inject constructor(
     }
 
     private suspend fun getF1TvSessions(event: F1TvSeasonEvent): List<F1TvSession> {
-        try {
-            val response = get(LIST_SESSIONS.format(getCurrentLocale(), event.meetingKey), sessionResponseJsonAdapter)
-            Log.d(TAG, "Fetched session ${event.id}")
+        val list: MutableList<F1TvSession> = ArrayList()
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:120]: Thanks.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 7]
```diff
@@ -60,14 +73,23 @@ class F1TvClient @Inject constructor(
...
     suspend fun getSessions(event: F1TvSeasonEvent, season: F1TvSeason): List<F1TvSession> {
-        return if (season.year.value < 2018) {
-            getSessionArchive(event, season)
-        } else {
-            getF1TvSessions(event)
+        return when {
+            season.year.value < 2018 -> {
+                getSessionArchive(event, season)
+            }
+            event.period.start < Instant.now() -> {
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:80]: I don't know where `archiveSortInstant` comes from or what it is supposed to contain. Can you elaborate?

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 8]
```diff
@@ -26,20 +30,29 @@ class F1TvClient @Inject constructor(
         private const val ROOT_URL = "https://f1tv.formula1.com"
 
         private const val GROUP_ID = 2 //TODO this might need to be migrated to the correct ONE
-        private const val LIST_SEASON = "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SEARCH/VOD/F1_TV_Pro_Monthly/$GROUP_ID?filter_objectSubtype=Meeting&filter_season=%s&filter_orderByFom=Y&maxResults=100"
-        private const val LIST_SESSIONS = "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SANDWICH/F1_TV_Pro_Monthly/$GROUP_ID?meetingId=%s&title=weekend-sessions"
-        private const val LIST_CHANNELS = "/2.0/R/%s/BIG_SCREEN_HLS/ALL/CONTENT/VIDEO/%s/F1_TV_Pro_Monthly/$GROUP_ID"
-        private const val PICTURE_URL = "https://ott.formula1.com/image-resizer/image/%s?w=384&h=384&o=L&q=HI"
+        private const val LIST_SEASON =
+            "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SEARCH/VOD/F1_TV_Pro_Monthly/$GROUP_ID?filter_objectSubtype=Meeting&filter_season=%s&filter_orderByFom=Y&maxResults=100"
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:34]: Yup.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt (hunk 1 above)
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:196]: Great.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 9]
```diff
@@ -117,12 +151,60 @@ class F1TvClient @Inject constructor(
...
-             * The pre seasons for example are not available to query
-             * on the regular api, for this fallback to the archive
-             */
             Log.d(TAG, "getF1TvSessions failed with ${e.message}")
-            //return getSessionArchive(event, season)
+            return listOf()
+        }
+    }
+
+    private suspend fun getFutureF1TvSessions(event: F1TvSeasonEvent): List<F1TvSession> {
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:159]: No, but when run in the emulator or on my tv, it feels just as quick as before.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 10]
```diff
@@ -26,20 +30,29 @@ class F1TvClient @Inject constructor(
...
-        private const val LIST_SESSIONS = "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SANDWICH/F1_TV_Pro_Monthly/$GROUP_ID?meetingId=%s&title=weekend-sessions"
-        private const val LIST_CHANNELS = "/2.0/R/%s/BIG_SCREEN_HLS/ALL/CONTENT/VIDEO/%s/F1_TV_Pro_Monthly/$GROUP_ID"
-        private const val PICTURE_URL = "https://ott.formula1.com/image-resizer/image/%s?w=384&h=384&o=L&q=HI"
+        private const val LIST_SEASON =
+            "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SEARCH/VOD/F1_TV_Pro_Monthly/$GROUP_ID?filter_objectSubtype=Meeting&filter_season=%s&filter_orderByFom=Y&maxResults=100"
+        private const val LIST_SESSIONS =
+            "/2.0/R/%s/BIG_SCREEN_HLS/ALL/PAGE/SANDWICH/F1_TV_Pro_Monthly/$GROUP_ID?meetingId=%s&title=weekend-sessions"
+        private const val LIST_FUTURE_SESSIONS =
+            "/2.0/R/%s/WEB_DASH/ALL/PAGE/1349/F1_TV_Pro_Monthly/$GROUP_ID"
+        private const val LIST_CHANNELS =
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt]: I have tried to use `BIG_SCREEN_HLS` as well and the data seems to be available now. But, I will have to wait until the next grand prix weekend to see if is showing future broadcasts as required.
I suggest to keep this pull request open until that time. When I have tested and confirmed `BIG_SCREEN_HLS` is also working, I will finish this pull request.

[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt]: @leonardoxh BIG_SCREEN_HLS also works, but I had to change to page 1350. See code changes.
It now works as desired. It shows the weekends schedule as of the first day of the weekend (so as of today for AZERBAIJAN).

**Comment by bashopman** [comment]
@leonardoxh Final result using BIG_SCREEN_HLS and all review feedback processed:
image

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (leonardoxh) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:73]: I think we have a class variable, but seeing it now if we use the same cached variable we can run into problems if the app is open for a long time. So this is good.

File: app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt [hunk 11]
```diff
@@ -117,12 +151,60 @@ class F1TvClient @Inject constructor(
...
-             * on the regular api, for this fallback to the archive
-             */
             Log.d(TAG, "getF1TvSessions failed with ${e.message}")
-            //return getSessionArchive(event, season)
+            return listOf()
+        }
+    }
+
+    private suspend fun getFutureF1TvSessions(event: F1TvSeasonEvent): List<F1TvSession> {
+        val dateTimeFormatter: DateTimeFormatter =
```
[the remark this answers is not in the document]
Comment (bashopman) [reply app/src/main/java/fr/groggy/racecontrol/tv/f1tv/F1TvClient.kt:150]: Android Studio cannot resolve `dateTimeFormatter`:
image

**Comment by bashopman** [comment]
@leonardoxh Happy with the result? Can we merge?

**Comment by leonardoxh** [comment]
Yup will merge it today

**Pull request merged**